2817: Add In-Pod Activity Log recorder in Audit JSON lines format

ngopalak-redhat opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

- If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Adds Audit JSON ability to security profiles operator. Outputs JSON lines for:
- User activity done inside "kubectl exec"
- Commands executed by the container itself

Usage is described in installation.md file of this PR.

Note:
- The feature currently only works if Seccomp profile can be enabled for this pod and the pod is not using privileged containers.
- Currently supports writing the JSON lines log at 1 minute interval to standard output. It can be made configurable later

#### Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.

Fixes #2817 

or

None
-->

#### Does this PR have test?
Yes. Its added in the PR itself and run on the Fedora and Ubuntu OS
<!--
If tests aren't applicable just write N/A.
-->

#### Special notes for your reviewer:
This is the first implementation for the feature. Once this is merged we can add features to make it more configurable. Sample output is available here: https://gist.github.com/ngopalak-redhat/f5552937e8d261493aa0bd8cb43640ee

#### Does this PR introduce a user-facing change?

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Add In-Pod Activity Log recorder in audit JSON lines format
```

saschagrunert commented:
/ok-to-test

ngopalak-redhat commented:
/retest

ngopalak-redhat commented:
/test e2e-seccomp-profile

ngopalak-redhat commented:
/test pull-security-profiles-operator-test-e2e

ngopalak-redhat commented:
/test pull-security-profiles-operator-test-e2e

ngopalak-redhat commented:
/review @BhargaviGudi

saschagrunert commented:
/retest

ccojocar commented on installation-usage.md:
````diff
@@ -847,6 +848,98 @@ Note that a security profile that is in use by existing pods cannot be
+   
+   To monitor the audit log tail:
+   ```shell
+   oc -n security-profiles-operator logs --since=1m --selector name=spod -c json-enricher --max-log-requests 6 -f
````
nit: kubectl

ccojocar commented on internal/pkg/manager/spod/bindata/spod.go:
```diff
@@ -615,6 +617,58 @@ semodule -i /opt/spo-profiles/selinuxrecording.cil
+						},
+						SecurityContext: &corev1.SecurityContext{
+							ReadOnlyRootFilesystem: &truly,
+							Privileged:             &truly,
```
Does this container need to run as `privileged`?

ngopalak-redhat replied:
Fixed and tested

ngopalak-redhat commented on installation-usage.md:
````diff
@@ -847,6 +848,98 @@ Note that a security profile that is in use by existing pods cannot be
+To start using this feature, you need to have the Security Profiles Operator installed in your Kubernetes cluster. 
+Once it's installed, you can enable the JSON log enricher with this command:
+```sh
+oc -n security-profiles-operator patch spod spod --type=merge -p '{"spec":{"enableJsonEnricher":true,"verbosity":0}}'
````
Thanks fixed

ccojocar commented:
@ngopalak-redhat It's a lgtm form my side. Thanks for contributing this. I can approve it if is ready to be merged.

haircommander commented on installation-usage.md:
```diff
@@ -847,6 +848,98 @@ Note that a security profile that is in use by existing pods cannot be
+- Tracking when someone uses commands like kubectl exec to get into a running container and run commands or scripts.
+- Monitoring activities in debug containers where users might run various tools.
+
+Currently, this feature is hardcoded to generate log for a process at 60 second interval. Also the JSON lines are output 
```
nit: `to generate logs for a process at an interval of 60 seconds`

haircommander commented on deploy/helm/values.yaml:
```diff
@@ -26,6 +26,7 @@ selinuxdImage:
 
 enableSelinux: false
 enableLogEnricher: false
+enableJsonEnricher: false
```
if there's going to be a configurable log interval period in the future, it almost makes sense to me to have this be a pointer to a jsonEnricher object that has an interval embedded. if the object is nil, no enrichment. if the interval is unset (0) then it defaults to 60 seconds, otherwise it uses the value of the interval

haircommander commented on internal/pkg/daemon/enricher/impl.go:
```diff
@@ -165,3 +172,41 @@ func (d *defaultImpl) Stat(name string) (os.FileInfo, error) {
+	defer func() {
+		cerr := file.Close()
+		if err == nil {
+			err = cerr
```
follow on: I like using a different name like `retErr` instead of `err` because weird scoping bugs can result in unexpected values. then, always set `err` and have `retErr` be set by the return statement, and reference it in the `defer`

ngopalak-redhat commented:
@haircommander / @ccojocar  I have addressed the review comments. Once this is merged, I will create a issue/PR for:
1. Making the audit log interval configurable.
2. Add an option to redirect the log output to a file rather than the standard output. Some observability systems like files instead of standard output
3. Raise a feature requirement to get human user information into the log output. Currently system user name is logged. This requires several changes which I'll explain in the feature.
All these are enhancements.

haircommander commented on api/spod/v1alpha1/spod_types.go:
```diff
@@ -276,6 +284,12 @@ type SPODSpec struct {
+	EnableJsonEnricher bool `json:"enableJsonEnricher,omitempty"`
+	// Defines options specific to the JsonEnricher
+	// functionality of the SecurityProfilesOperator
+	JsonEnricherOpts JsonEnricherOptions `json:"jsonEnricherOptions,omitempty"`
```
IMO this should be a pointer to JsonEnricherOptions and then we don't need the enablement bool
I also think JSON would be better? though that may have it serialize oddly?

ngopalak-redhat replied:
- I made `JsonEnricherOpts` as a pointer
- `EnableJsonEnricher` bool is required to launch this container. Hence its retained
- "JSON" in the variable name is not camel cased and from reading several docs its recommended to keep CamelCase variables. Esp when serialization is involved.

ngopalak-redhat replied:
@haircommander `EnableJsonEnricher` works similar to `EnableLogEnricher`. There is a e2e test that's using it: `https://github.com/kubernetes-sigs/security-profiles-operator/blob/ecdd64fc47e88dd5fe0faf5937517c2b9d34dd83/test/suite_test.go#L782` . Hence its required

ccojocar commented:
/lgtm
